### PR TITLE
ISSUE-853 [FILTER] Forward action

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterGetMethodContract.scala
Comment (Arsnael) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterGetMethodContract.scala:676]: indent

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FilterGetMethod.scala
Comment (vttranlina) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FilterGetMethod.scala:82]: For debug?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FilterGet.scala
Comment (quantranhong1999) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FilterGet.scala:95]:
```scala
import scala.jdk.OptionConverters._

forwardOptional.toScala
    .map(...)
```

is more functional.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FilterSet.scala
Comment (quantranhong1999) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FilterSet.scala:83]:
```scala
import scala.jdk.OptionConverters._

forwardOption.toJava
    .map(...)
```

less verbose IMO.

**Review by quantranhong1999 — COMMENTED** [review]
Likely missing `Update tmail backend JMAP doc for filter`

**Comment by Arsnael** [comment]
Can rebase

**Comment by chibenwa** [comment]
@hungphan227 do you plan of handling comments here?

**Comment by hungphan227** [comment]
I will update this PR after the forward loop PR is merged

**Comment by chibenwa** [comment]
> I will update this PR after the forward loop PR is merged

Why wait?

**Comment by hungphan227** [comment]
anyway, I need some time to update this

**Comment by hungphan227** [comment]
> Likely missing `Update tmail backend JMAP doc for filter`

doc added

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterSetMethodContract.scala
Comment (quantranhong1999) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterSetMethodContract.scala]: Q: Any issue/bug keeping this?

**Review by quantranhong1999 — APPROVED** [review]

**Comment by Arsnael** [comment]
```
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile (scala-compile-first) on project jmap-extensions: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile failed: Compilation failed: InterfaceCompileFailed -> [Help 1]
```

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FilterSet.scala
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FilterSet.scala:83]:
```
forwardOption.map(forward => Optional.of(Rule.Action.Forward.of(
    forward.addresses.map(address => new MailAddress(address.string)).asJava,
    forward.keepACopy.value)))
  .getOrElse(Optional.empty())
```
?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FilterGet.scala
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/FilterGet.scala:91]:
```
  .map(actionForward => FilterForward(
    actionForward.getAddresses.asScala.toList.map(mailAddress => MailAddress(mailAddress.asString())),
    KeepACopy(actionForward.isKeepACopy)))
```

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FilterGetMethod.scala
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/FilterGetMethod.scala:81]:
```
  .map(rulesWithVersion => FilterWithVersion(
    Filter("singleton", rulesWithVersion.getRules.asScala.toList.map(rule => Rule.fromJava(rule, mailboxIdFactory))),
    rulesWithVersion.getVersion))
```

**Review by Arsnael — COMMENTED** [review]
More like readability comments but ok to me overall

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterSetMethodContract.scala
Comment (hungphan227) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterSetMethodContract.scala]: No. We can have it or not. The results are the same

**Comment by hungphan227** [comment]
> ```
> [ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile (scala-compile-first) on project jmap-extensions: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile failed: Compilation failed: InterfaceCompileFailed -> [Help 1]
> ```

There is some change in PR filter loop PR that impact this PR. Maybe I have to wait for the filter loop PR to be merged

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterSetMethodContract.scala
Comment (hungphan227) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraFilterSetMethodContract.scala]: I have updated code in james so that in case withKeywords field in json file does not exists, the empty list will be created for withKeywords in the Rule object

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**
